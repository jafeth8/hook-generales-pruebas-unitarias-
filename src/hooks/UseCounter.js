import { useState } from "react"

export const UseCounter = (initialState=0) => {

    const [counter, setCounter] = useState(initialState)

    const incrementar=() => {
        setCounter(counter+1)
    }

    const resetear=() => {
        setCounter(initialState)
    }

    const decrementar=() => {
        setCounter(counter-1)        
    }

    return {counter,incrementar,resetear,decrementar}
    
}
