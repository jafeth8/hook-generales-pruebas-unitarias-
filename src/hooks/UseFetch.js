import { useEffect, useRef, useState } from 'react'

export const UseFetch = (url) => {

    const [state, setState] = useState({data:null,loading:true,error:null});

    const isMouted = useRef(true);
    
    useEffect(() => {
        
        return () => {
            isMouted.current=false;
        }
    }, [])

    useEffect(()=>{
        
        //setState({data:null,loading:true,error:null}); // (instruccion para mostrar loading en <MultipleCustomHooks/>)
        setTimeout(() => {

            
            if(isMouted.current){
                fetch(url)
                    .then(resp=>resp.json())
                    .then((data) =>{
                    setState({data,loading:false,error:null})
                });
            }else{
                console.log("no se puede cambiar el estado de un componente no montado")
            }            

        }, 3000);

    },[url])

    return state;
}
