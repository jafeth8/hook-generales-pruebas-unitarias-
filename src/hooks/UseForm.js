import {useState} from 'react'

export const UseForm = (initialState={}) => {

    const [values, setValues] = useState(initialState);

    const reset=()=>{
        setValues(initialState);
    }

    const handleChange=({target})=>{
        //console.log(e.target.name)
        //console.log(e.target.value)
        setValues({
            ...values,
            [target.name]:target.value
        })
    }

    return [values, handleChange,reset]
    
}
