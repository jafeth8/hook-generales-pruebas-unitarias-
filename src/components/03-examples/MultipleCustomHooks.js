import React from 'react'
import { UseFetch } from '../../hooks/UseFetch'
import {UseCounter} from '../../hooks/UseCounter'
import "../02-useEffect/Effects.css"
export const MultipleCustomHooks = () => {

    const {counter,incrementar} =UseCounter(1);

   
    //const state=UseFetch(`https://www.breakingbadapi.com/api/quotes/1`);
    //console.log(state);
    const {loading,data}=UseFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    console.log(data);
    const {quote,author}= !!data && data[0];
    //console.log(quote,author);

    return (
        <div>
            <h1> BreakingBad Quotes! </h1>
            <hr/>
            {
                loading 
                ?
                    (
                        <div className="alert alert-info text-center">
                            loading
                        </div>
                    )

                :

                    (
                        <blockquote className="blockquote text-end">
                            <p className="mb-3">{quote}</p>
                            <footer className="blockquote-footer">{author}</footer>
                        </blockquote>
                    )

                
            }

            <button className="btn btn-primary" onClick={incrementar}> Siguiente Quote XD</button>

        </div>
    )
}
