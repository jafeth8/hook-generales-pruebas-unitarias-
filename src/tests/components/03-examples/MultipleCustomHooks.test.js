import {shallow} from 'enzyme';
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { UseCounter } from '../../../hooks/UseCounter';
import { UseFetch } from '../../../hooks/UseFetch';

jest.mock('../../../hooks/UseFetch')

jest.mock('../../../hooks/UseCounter');



describe('pruebas en componente <MultipleCustomHooks/>', () => {

    beforeEach(() => { //demostracion de que se pueden usar mas de 2 "mocks"
        UseCounter.mockReturnValue({
            counter:1,
            incrementar:()=>{},
        })
    })


    test('mostrar correctamente el componente', () => {

        UseFetch.mockReturnValue({
            data:null,loading:true,error:null
        })


        const wrapper =shallow(<MultipleCustomHooks/>);
        expect(wrapper).toMatchSnapshot();

    })

    test('debe mostrar la informacion', () => {

        UseFetch.mockReturnValue({
            data:[
                {author:'Romeo', quote:'confia en el camino que elegiste'}
            ],
            loading:false,
            error:null
        })

      
        const wrapper =shallow(<MultipleCustomHooks/>);
        //console.log(wrapper.html());

        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('blockquote').exists()).toBe(true);
        expect(wrapper.find('.mb-3').text().trim()).toBe('confia en el camino que elegiste');
        expect(wrapper.find('footer').text().trim()).toBe('Romeo');
        
    })
    
    
})
