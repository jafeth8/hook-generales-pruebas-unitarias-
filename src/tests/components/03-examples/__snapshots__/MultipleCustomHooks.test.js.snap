// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en componente <MultipleCustomHooks/> mostrar correctamente el componente 1`] = `
<div>
  <h1>
     BreakingBad Quotes! 
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    loading
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
     Siguiente Quote XD
  </button>
</div>
`;
